pkg/controllers/disruption/emptiness.go

```diff
@@ -18,74 +18,120 @@ package disruption
-				return true
-			}
-			return false
-		}),
-	}, scheduling.Results{}, nil
+//
+//nolint:gocyclo
+func (e *Emptiness) ComputeCommand(ctx context.Context, disruptionBudgetMapping map[string]map[v1.DisruptionReason]int, candidates ...*Candidate) (Command, scheduling.Results, error) {
+	if e.IsConsolidated() {
+		return Command{}, scheduling.Results{}, nil
+	}
+	candidates = e.sortCandidates(candidates)
+
+	empty := make([]*Candidate, 0, len(candidates))
+	constrainedByBudgets := false
+	for _, candidate := range candidates {
+		if len(candidate.reschedulablePods) > 0 {
+			continue
+		}
+		if disruptionBudgetMapping[candidate.nodePool.Name][v1.DisruptionReasonEmpty] == 0 {
+			// set constrainedByBudgets to true if any node was a candidate but was constrained by a budget
+			constrainedByBudgets = true
+			continue
+		}
+		// If there's disruptions allowed for the candidate's nodepool,
+		// add it to the list of candidates, and decrement the budget.
+		empty = append(empty, candidate)
+		disruptionBudgetMapping[candidate.nodePool.Name][v1.DisruptionReasonEmpty]--
+	}
+	// none empty, so do nothing
+	if len(empty) == 0 {
+		// if there are no candidates, but a nodepool had a fully blocking budget,
+		// don't mark the cluster as consolidated, as it's possible this nodepool
+		// should be consolidated the next time we try to disrupt.
+		if !constrainedByBudgets {
+			e.markConsolidated()
+		}
+		return Command{}, scheduling.Results{}, nil
+	}
+
+	cmd := Command{
+		candidates: empty,
+	}
+
+	// Empty Node Consolidation doesn't use Validation as we get to take advantage of cluster.IsNodeNominated.  This
+	// lets us avoid a scheduling simulation (which is performed periodically while pending pods exist and drives
+	// cluster.IsNodeNominated already).
+	select {
+	case <-ctx.Done():
+		return Command{}, scheduling.Results{}, errors.New("interrupted")
+	case <-e.clock.After(consolidationTTL):
+	}
+
+	v := NewValidation(e.clock, e.cluster, e.kubeClient, e.provisioner, e.cloudProvider, e.recorder, e.queue, v1.DisruptionReasonEmpty)
+	validatedCandidates, err := v.ValidateCandidates(ctx, cmd.candidates...)
+	if err != nil {
+		if IsValidationError(err) {
+			log.FromContext(ctx).V(1).Info(fmt.Sprintf("abandoning empty node consolidation attempt due to pod churn, command is no longer valid, %s", cmd))
+			return Command{}, scheduling.Results{}, nil
+		}
+		return Command{}, scheduling.Results{}, err
+	}
+
+	// TODO (jmdeal@): better encapsulate within validation
+	if lo.ContainsBy(validatedCandidates, func(c *Candidate) bool {
+		return len(c.reschedulablePods) != 0
+	}) {
+		log.FromContext(ctx).V(1).Info(fmt.Sprintf("abandoning empty node consolidation attempt due to pod churn, command is no longer valid, %s", cmd))
+		return Command{}, scheduling.Results{}, nil
+	}
+
+	return cmd, scheduling.Results{}, nil
 }
 
 func (e *Emptiness) Type() string {
-	return metrics.EmptinessReason
+	return metrics.ConsolidationReason
```

Comment:
I think this should actually just be the ConsolidationReason, since that's what limits this disruption method. If we call this emptiness, we should call the others underutilized. It's more clear IMO to name them all consolidation, and then make the consolidationTypes differentiate them.

Reply:
So are you thinking we should name the Multi and Single Node consolidations Underutilization? At a core, the method we're doing is consolidation, empty or underutilized would be a consolidationType, which is how I've modeled it. I personally think it makes more sense this way.

Reply:
> empty or underutilized would be a consolidationType

We should discuss this more offline since it sounds like we don't agree, but my thought is that if we are going to call the reason of the budgets: "Empty" and "Underutilized"; then, we should call the reason for the `nodeclaims_disrupted_total` by the same name. In terms of the method type, I could be convinced that it's consolidation, but I also don't know if you don't just model this with a reason label as well